File: 1646-get-maximum-in-generated-array/1646-get-maximum-in-generated-array.py
class Solution:
    def getMaximumGenerated(self, n: int) -> int:
        if n == 0:
            return 0
        if n == 1:
            return 1
        dp = [-1]*(n+1)
        dp[0] = 0
        dp[1] = 1
        maxi = -2
        for i in range(2,n+1):
            if i%2==0:
                dp[i] = dp[i//2]
            else:
                dp[i] = dp[(i-1)//2] + dp[(i-1)//2+1]
            maxi = max(dp[i],maxi)
            
        return maxi
        
            
        

File: 1646-get-maximum-in-generated-array/1646-get-maximum-in-generated-array.cpp
class Solution {
public:
    int getMaximumGenerated(int n) {
        if(n<2) return n;
        vector<int> nums(n+2,0);
        nums[1]=1;
        int j = 1;
        int maxi = 0;
        while (2*j<=n) {
            nums[2*j] = nums[j];
            nums[2*j+1] = nums[j]+nums[j+1];
            j++;
            
        }
        for (int k=0;k<=n;k++)
            maxi = max(nums[k],maxi);
        return maxi;
        
    }
};